modules/mcc/include/opencv2/mcc/ccm.hpp

```diff
@@ -0,0 +1,545 @@
+        Color color(converted.reshape(3, 0), *(cs.l));
+        return calc_loss_(color);
+    }
+
+    /* *\ brief Fitting ccm if distance function is associated with CIE Lab color space.
+        *        see details in https://github.com/opencv/opencv/blob/master/modules/core/include/opencv2/core/optim.hpp
+        *        Set terminal criteria for solver is possible.
+    */
+    void fitting(void)
+    {
+        cv::Ptr<DownhillSolver> solver = cv::DownhillSolver::create();
+        cv::Ptr<LossFunction> ptr_F(new LossFunction(this));
+        solver->setFunction(ptr_F);
+        cv::Mat reshapeccm = ccm0.clone().reshape(0, 1);
+        cv::Mat step = cv::Mat::ones(reshapeccm.size(), CV_64F);
+        solver->setInitStep(step);
+        TermCriteria termcrit = TermCriteria(TermCriteria::MAX_ITER + TermCriteria::EPS, max_count, epsilon);
+        solver->setTermCriteria(termcrit);
+        double res = solver->minimize(reshapeccm);
+        ccm = reshapeccm.reshape(0, shape/3);
+        loss = pow((res / masked_len), 0.5);
+        std::cout << " ccm " << ccm << std::endl;
+        std::cout << " loss " << loss << std::endl;
+    };
+
+    /* *\ brief Infer using fitting ccm.
+        *\ param img the input image, type of cv::Mat.
+        *\ return the output array, type of cv::Mat.
+    */
+    cv::Mat infer(const cv::Mat& img, bool islinear = false)
+    {
+        if (!ccm.data)
+        {
+            throw "No CCM values!";
+        }
+        cv::Mat img_lin = linear->linearize(img);
+        cv::Mat img_ccm(img_lin.size(), img_lin.type());
+        cv::Mat ccm_ = ccm.reshape(0, shape / 3);
+        img_ccm = multiple(prepare(img_lin), ccm_);
+        if (islinear == true)
+        {
+            return img_ccm;
+        }
+        return cs.fromL(img_ccm);
+    };
+
+    /* *\ brief Infer image and output as an BGR image with uint8 type.
+        *        mainly for test or debug.
+        *        input size and output size should be 255.
+        *\ param imgfile path name of image to infer.
+        *\ param islinear if linearize or not.
+        *\ return the output array, type of cv::Mat.
+    */
+    cv::Mat inferImage(std::string imgfile, bool islinear = false)
+    {
+        const int inp_size = 255;
+        const int out_size = 255;
+        cv::Mat img = imread(imgfile);
+        cv::Mat img_;
+        cvtColor(img, img_, COLOR_BGR2RGB);
+        img_.convertTo(img_, CV_64F);
+        img_ = img_ / inp_size;
+        cv::Mat out = this->infer(img_, islinear);
+        cv::Mat out_ = out * out_size;
+        out_.convertTo(out_, CV_8UC3);
+        cv::Mat img_out = min(max(out_, 0), out_size);
+        cv::Mat out_img;
+        cvtColor(img_out, out_img, COLOR_RGB2BGR);
+        return out_img;
+    };
+
+    // void report() {
+    //     std::cout << "CCM0: " << ccm0 << std::endl;
+    //     std::cout << "CCM: " << ccm << std::endl;
+    //     std::cout << "Loss: " << loss << std::endl;
+    // }
+
+private:
```

Comment:
the public classes should not contain private methods and members. Either use abstract class to declare interface with separate "implementation" class (https://github.com/opencv/opencv/wiki/Coding_Style_Guide#high-level-c-interface-algorithms) or use `private: class Impl; Ptr<Impl> pImpl;` technique (https://en.cppreference.com/w/cpp/language/pimpl)